src/Spectre.libGaussianMixtureModelling/ExpectationMaximization.h

```diff
@@ -74,77 +74,69 @@ static inline DataType CalculateDelta(GaussianMixtureModel& oldModel,
-        : m_Spectrum(std::move(spectrum)), m_Components(components)
-          , m_AffilationMatrix((unsigned)components.size(), (unsigned)spectrum.mzs.size())
-          , m_Expectation(spectrum, m_AffilationMatrix, m_Components)
-          , m_Maximization(spectrum, m_AffilationMatrix, m_Components)
+    /// <param name="spectrum">Spectrum to be approximated.</param>
+    /// <param name="maxNumOfComponent">Number of Gaussian components
+    /// that build up the approximation.</param>
+    ExpectationMaximization(SpectrumView spectrum,
+                            unsigned maxNumOfComponents)
+          : m_AffilationMatrix(maxNumOfComponents,(unsigned)spectrum.mzs.size())
+          , m_Expectation(spectrum)
+          , m_Maximization(spectrum)
     {
     }
 
     /// <summary>
     /// Performs a full algorithm run. Terminates when changes in weights and
     /// heights are below threshold defined by epsilon.
     /// </summary>
-    /// <param name="minStd">Minimal standard deviation a component can have.
+    /// <param name="spectrum">Data to approximate.</param>
+    /// <param name="components">Base components to improve.</param>
+    /// <param name="minStd">
+    /// Minimal standard deviation a component can have.
+    /// </param>
+    /// <param name="epsilon">
+    /// Threshold determining max difference in parameters between consecutive
+    /// iterations.
     /// </param>
-    /// <param name="epsilon">Threshold determining max difference in
-    /// parameters between consecutive iterations.</param>
     /// <returns>
     /// Gaussian Mixture Model containing all the components with their
     /// appropriate parameters.
     /// </returns>
-    GaussianMixtureModel EstimateGmm(DataType minStd,
-        DataType epsilon)
+    GaussianMixtureModel EstimateGmm(SpectrumView spectrum,
+        GaussianMixtureModel& components, DataType minStd, DataType epsilon)
     {
-        SetMinStds(m_Components, minStd);
-        GaussianMixtureModel oldModel(m_Components.size());
+        SetMinStds(components, minStd);
+        GaussianMixtureModel oldModel(components.size());
 
         DataType delta; // How much did the last iteration change the results.
         do
         {
-            FilterLowHeightComponents(m_Components);
-            oldModel.assign(m_Components.begin(), m_Components.end());
-            Expectation();
-            Maximization();
-            delta = CalculateDelta(oldModel, m_Components);
+            FilterLowHeightComponents(components);
+            oldModel.assign(components.begin(), components.end());
+            m_Expectation.Expectation(m_AffilationMatrix, spectrum, components);
+            m_Maximization.Maximization(m_AffilationMatrix, spectrum,
+                components);
+            delta = CalculateDelta(oldModel, components);
         } while (delta > epsilon);
 
-        return m_Components; // Consider sorting before return
+        return components;
```

Comment:
Yes I am aware of it, but to comply with the unwritten standard of native-algorithms libs, I decided to return it anyway, because @gmrukwa would complain.